lib/plugins/output-filter/kubernetes-enrichment.js

```diff
@@ -0,0 +1,212 @@
+  var index = podCache[key].metadata.annotations['sematext.com/logs-token']
+  // get comma separated list of fields from pod annotation
+  if (index) {
+    data._index = index
+  } else {
+    podCache[key].stLogsTokenSet = false
+  }
+}
+/*
+
+function updatePodUid (key, data) {
+  if (data.kubernetes && podCache[key]) {
+    data.kubernetes.pod.uid = podCache[key].metadata.uid
+  } else {
+    podCache[key].stLogsTokenSet = false
+  }
+}
+
+*/
+
+function replaceDockerImageName (key, data) {
+  if (data.kubernetes && data.container && podCache[key].stImageCache) {
+    var containerName = data.kubernetes.pod.container.name
+    var imageInfo = podCache[key].stImageCache[data.kubernetes.pod.name + containerName]
+    if (!imageInfo) {
+      return
+    }
+    if (data.container.image.name === 'sha256') {
+      data.container.image.digest = 'sha256:' + data.container.image.tag
+    }
+    data.container.image.name = imageInfo.name
+    data.container.image.tag = imageInfo.tag
+    // data.container.image.registry = imageInfo.version
+    // data.container.image.name = imageInfo.plainImageName
+  } else {
+    podCache[key].stLogsTokenSet = false
+  }
+}
+
+function processAnnotations (data, context) {
+  var key = data.kubernetes.namespace + '/' + data.kubernetes.pod.name
+  if (podCache[key] && podCache[key].metadata) {
+    checkLogsEnabled(key, data, context)
+    if (data.stLogEnabled === false) {
+      // logs will be droped, so no need for other processing
+      return
+    }
+    // updatePodUid (key, data)
+    replaceDockerImageName(key, data)
+    removeFields(key, data)
+    addLogsIndex(key, data)
+  }
+}
+
+function enrichLogs (context, config, eventEmitter, data, callback) {
+  if (data.kubernetes && data.kubernetes.pod) {
+    if (podCache[data.kubernetes.namespace + '/' + data.kubernetes.pod.name]) {
+      processAnnotations(data, context)
+      if (data.stLogEnabled === false) {
+        // allow input plugins to close the input stream, e.g. input/docker/docker.js
+        eventEmitter.emit('dropLogsRequest', context, data)
+        if (config.debug == true) { 
+          console.log('logs dropped ', data.kubernetes.namespace + '/' + data.kubernetes.pod.name, data.message)
+        }
+        return callback()
+      } else {
+        return callback(null, data)
+      }
+    } else {
+      getPodSpec(
+        config,
+        data.kubernetes.namespace,
+        data.kubernetes.pod.name,
+        function (err, pod) {
+          if (err) {
+            consoleLogger.log(err.message)
+            return callback(null, data)
+          }
+          podCache[data.kubernetes.namespace + '/' + data.kubernetes.pod.name] = pod
+          // create hashtable with containerName -> imageInformation
+          pod.stImageCache = {}
+          if (pod.spec && pod.spec.containers) {
+            var podContainers = pod.spec.containers
+            for (var i = 0; i < podContainers.length; i++) {
```

Comment:
Code with a lot of `for` loops is hard to refactor and read. If you start using functional JavaScript more, with higher-order functions and array methods like `.forEach`, `.map`, etc. you can make the code more linear and more immutable by using `const` and creating new variables instead of re-assigning old ones. This can help reduce bugs and potential memory leaks.

Reply:
for loops have a better performance than `forEach(function(){})`. Imagine LA processes thousands of events per second so even little differences in performance sum up.